library/psa_crypto.c

```diff
@@ -1342,6 +1365,217 @@ psa_status_t psa_export_key( mbedtls_svc_key_id_t key,
+            break;
+        case 512:
+            dhm_P = dhm_P_4096;
+            dhm_G = dhm_G_4096;
+            dhm_size_P = sizeof( dhm_P_4096 );
+            dhm_size_G = sizeof( dhm_G_4096 );
+            break;
+        case 768:
+            dhm_P = dhm_P_6144;
+            dhm_G = dhm_G_6144;
+            dhm_size_P = sizeof( dhm_P_6144 );
+            dhm_size_G = sizeof( dhm_G_6144 );
+            break;
+        case 1024:
+            dhm_P = dhm_P_8192;
+            dhm_G = dhm_G_8192;
+            dhm_size_P = sizeof( dhm_P_8192 );
+            dhm_size_G = sizeof( dhm_G_8192 );
+            break;
+        default:
+            return( PSA_ERROR_INVALID_ARGUMENT );
+            break;
+    }
+
+    if( P != NULL )
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( P, dhm_P,
+                                                  dhm_size_P ) );
+    if( G != NULL )
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( G, dhm_G,
+                                                  dhm_size_G ) );
+
+cleanup:
+    if( ret != 0 )
+        return( mbedtls_to_psa_error( ret ) );
+
+    return( PSA_SUCCESS );
+}
+
+static psa_status_t psa_key_agreement_ffdh( const uint8_t *peer_key,
+                                            size_t peer_key_length,
+                                            const uint8_t *our_key,
+                                            size_t our_key_length,
+                                            uint8_t *shared_secret,
+                                            size_t shared_secret_size,
+                                            size_t *shared_secret_length )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    mbedtls_mpi P, G, X, GX, GY, K;
+    const size_t calculated_shared_secret_size = peer_key_length;
+
+    if(peer_key_length != our_key_length ||
+       calculated_shared_secret_size > shared_secret_size )
+    {
+        return(PSA_ERROR_INVALID_ARGUMENT);
+    }
+
+    mbedtls_mpi_init( &P ); mbedtls_mpi_init( &G );
+    mbedtls_mpi_init( &X ); mbedtls_mpi_init( &GX );
+    mbedtls_mpi_init( &GY ); mbedtls_mpi_init( &K );
+
+    status = psa_ffdh_set_prime_generator( peer_key_length, &P, &G );
+
+    if( status == PSA_SUCCESS )
+    {
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( &X, our_key,
+                                                our_key_length ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( &GY, peer_key,
+                                                peer_key_length ) );
+
+        /* Calculate our public key: GX = G^X mod P */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_exp_mod( &GX, &G, &X, &P , NULL ) );
+
+        /* Calculate shared secret public key: K = G^(XY) mod P */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_exp_mod( &K, &GY, &X, &P, NULL ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( &K, shared_secret,
+                                                   calculated_shared_secret_size ) );
```

Comment:
I assume that `calculated_shared_secret_size` variable can be dropped and here `shared_secret_size` should be used instead. `*shared_secret_length` should be set using `mbedtls_mpi_size( &K );`

Unfortunately in this case test fails when `shared_secret_size` is greater than required:
https://github.com/Mbed-TLS/mbedtls/blob/6dc7cdfdd18eeb9171e4a8ef5f08af8e2a4043d2/tests/suites/test_suite_psa_crypto.function#L7590-L7597

I don't know why in such case different key is stored into the buffer. Zero byte is at the begging, length is the same and last byte is missing:

```
 output    : 00718ab2b5da3bc6e7767a98fb2c172bd74003fae2acffbc9a53d9b358401c1c748da36cab277e9397bc5eeec3010321d0f882d959eb097adddc99745526b213e30dc0df9fb1e4cd3fc27bfb1d6e89c715373439a66b9a13aa1334c84799827c17be1c36c1bc02fe60ea698da790fe4d2af710a435a1aae7fb11cd2a90a17ad87dde4f154b325dc47d8ea107a29d10a3bfa17149a1f9e8a1f7b680bfdca90fb0913c0b681670d904de49d7d000d24060330d4d2e4a2381d78c49e272d313174218561ceeb37e2ef824905d0fa42d13d49a73018411aeb749f7f4fc765bdc6db58bcebd995d4c949b0061f20759e1263d8f9ba3fd56afda07c178997256bb7d52

 exp output: 718ab2b5da3bc6e7767a98fb2c172bd74003fae2acffbc9a53d9b358401c1c748da36cab277e9397bc5eeec3010321d0f882d959eb097adddc99745526b213e30dc0df9fb1e4cd3fc27bfb1d6e89c715373439a66b9a13aa1334c84799827c17be1c36c1bc02fe60ea698da790fe4d2af710a435a1aae7fb11cd2a90a17ad87dde4f154b325dc47d8ea107a29d10a3bfa17149a1f9e8a1f7b680bfdca90fb0913c0b681670d904de49d7d000d24060330d4d2e4a2381d78c49e272d313174218561ceeb37e2ef824905d0fa42d13d49a73018411aeb749f7f4fc765bdc6db58bcebd995d4c949b0061f20759e1263d8f9ba3fd56afda07c178997256bb7d5230
```

Reply:
I think i found the problem. When you use mbedtls_mpi_write_binary(), it saves data using big endian. And when data is read from buffer, it read as little endian.

Using mbedtls_mpi_write_binary_le() ( it writes data as little endian ) in psa_crypto.c, it solves problem. But in order to pass test, expected result need to be written in little endian.

My test passed when i did following things:
- change binary to binary_le()
- change data files.

After this, it was okay to use shared_secret_size and use mbedtls_mpi_size( &K ) to load into shared_secret length.

Reply:
The shared secret for [PSA_ALG_FFDH](https://armmbed.github.io/mbed-crypto/html/api/ops/ka.html?highlight=psa_alg_ffdh#c.PSA_ALG_FFDH) should be represented in big-endian order.

Reply:
Ah right. So current way of calculating shared_secret_size is best way because calculated_shared_secret_size and shared_secret_size is not necessarily same?